From 288547510a07e47e756da729c35bb6bcce566dc0 Mon Sep 17 00:00:00 2001
From: Allan Sandfeld Jensen <allan.jensen@digia.com>
Date: Wed, 4 Dec 2013 12:51:28 +0100
Subject: [PATCH] Avoid duplicate emulated mouse events with XInput2

When using a touch screen on a Linux machine, we receive both touch-events
and emulated mouse events from XInput, on top of that we synthesize mouse-
events ourselves for the touch events.

This patch grabs the touch device for touch events whenever it processes
a touch-begin thereby avoiding XInput from synthesizing mouse events.

Task-number: QTBUG-35157
Change-Id: I5849d5841be236d6719cd080af2e9e39eb9cdd84
Reviewed-by: Laszlo Agocs <laszlo.agocs@digia.com>
(cherry picked from commit 2c65b78b400ec27e6e559829b9a970dca2df6669)

Conflicts:
	src/plugins/platforms/xcb/qxcbconnection_xi2.cpp
---
 src/plugins/platforms/xcb/qxcbconnection_xi2.cpp | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/src/plugins/platforms/xcb/qxcbconnection_xi2.cpp b/src/plugins/platforms/xcb/qxcbconnection_xi2.cpp
index 991c82e..0708f70 100644
--- a/src/plugins/platforms/xcb/qxcbconnection_xi2.cpp
+++ b/src/plugins/platforms/xcb/qxcbconnection_xi2.cpp
@@ -323,6 +323,22 @@ void QXcbConnection::xi2HandleEvent(xcb_ge_event_t *event)
                     " area " << touchPoint.area << " pressure " << touchPoint.pressure;
 #endif
                 QWindowSystemInterface::handleTouchEvent(platformWindow->window(), xiEvent->time, dev->qtTouchDevice, m_touchPoints.values());
+                if (has_touch_without_mouse_emulation) {
+                    // We need to grab the touch event to prevent mouse emulation.
+                    if (xiEvent->evtype == XI_TouchBegin) {
+                        XIEventMask xieventmask;
+                        unsigned int bitMask = 0;
+                        unsigned char *xiBitMask = reinterpret_cast<unsigned char *>(&bitMask);
+                        xieventmask.deviceid = xiEvent->deviceid;
+                        xieventmask.mask = xiBitMask;
+                        xieventmask.mask_len = sizeof(bitMask);
+                        bitMask |= XI_TouchBeginMask;
+                        bitMask |= XI_TouchUpdateMask;
+                        bitMask |= XI_TouchEndMask;
+                        XIGrabDevice(static_cast<Display *>(m_xlib_display), xiEvent->deviceid, platformWindow->winId(), xiEvent->time, None, GrabModeAsync, GrabModeAsync, true, &xieventmask);
+                    } else if (xiEvent->evtype == XI_TouchEnd)
+                        XIUngrabDevice(static_cast<Display *>(m_xlib_display), xiEvent->deviceid, xiEvent->time);
+                }
                 // If a touchpoint was released, we can forget it, because the ID won't be reused.
                 if (touchPoint.state == Qt::TouchPointReleased)
                     m_touchPoints.remove(touchPoint.id);
-- 
1.9.1

